tests/integration/test_backups.py

```diff
@@ -0,0 +1,329 @@
+
+import json
+import logging
+import os
+from pathlib import Path
+
+import boto3
+import pytest
+from pytest_operator.plugin import OpsTest
+
+from .helpers import (
+    execute_queries_on_unit,
+    get_primary_unit,
+    get_server_config_credentials,
+    get_unit_address,
+    rotate_credentials,
+    scale_application,
+)
+from .high_availability.high_availability_helpers import (
+    deploy_and_scale_mysql,
+    insert_data_into_mysql_and_validate_replication,
+)
+
+logger = logging.getLogger(__name__)
+
+CLOUD_CONFIGS = {
+    "aws": {
+        "endpoint": "https://s3.amazonaws.com",
+        "bucket": "canonical-mysql",
+        "path": "test",
+        "region": "us-east-1",
+    },
+    "gcp": {
+        "endpoint": "https://storage.googleapis.com",
+        "bucket": "data-charms-testing",
+        "path": "mysql-k8s",
+        "region": "",
+    },
+}
+CLOUD_CREDENTIALS = {
+    "aws": {
+        "access-key": os.environ["AWS_ACCESS_KEY"],
+        "secret-key": os.environ["AWS_SECRET_KEY"],
+    },
+    "gcp": {
+        "access-key": os.environ["GCP_ACCESS_KEY"],
+        "secret-key": os.environ["GCP_SECRET_KEY"],
+    },
+}
+S3_INTEGRATOR = "s3-integrator"
+TIMEOUT = 10 * 60
+CLUSTER_ADMIN_PASSWORD = "clusteradminpassword"
+SERVER_CONFIG_PASSWORD = "serverconfigpassword"
+ROOT_PASSWORD = "rootpassword"
+DATABASE_NAME = "backup-database"
+TABLE_NAME = "backup-table"
+
+backups_by_cloud = {}
+value_before_backup, value_after_backup = None, None
+
+
+@pytest.fixture(scope="session", autouse=True)
+def clean_backups_from_buckets() -> None:
+    """Teardown to clean up created backups from clouds."""
+    yield
+
+    logger.info("Cleaning backups from cloud buckets")
+    for cloud_name, config in CLOUD_CONFIGS.items():
+        backup = backups_by_cloud[cloud_name]
+
+        if not backup:
+            continue
+
+        session = boto3.session.Session(
+            aws_access_key_id=CLOUD_CREDENTIALS[cloud_name]["access-key"],
+            aws_secret_access_key=CLOUD_CREDENTIALS[cloud_name]["secret-key"],
+            region_name=config["region"],
+        )
+        s3 = session.resource("s3", endpoint_url=config["endpoint"])
+        bucket = s3.Bucket(config["bucket"])
+
+        # GCS doesn't support batch delete operation, so delete the objects one by one
+        backup_path = str(Path(config["path"]) / backups_by_cloud[cloud_name])
+        for bucket_object in bucket.objects.filter(Prefix=backup_path):
+            bucket_object.delete()
+
+
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Simple test to ensure that the mysql charm gets deployed."""
+    mysql_application_name = await deploy_and_scale_mysql(ops_test)
+
+    mysql_unit = ops_test.model.units[f"{mysql_application_name}/0"]
+    primary_mysql = await get_primary_unit(ops_test, mysql_unit, mysql_application_name)
+
+    logger.info("Rotating all mysql credentials")
+
+    await rotate_credentials(
```

Comment:
why rotate credentials?

Reply:
when we create a backup, the credentials are also backed up. when we restore to a new cluster, the credentials need to be reset (which will be done in a separate ticket) or rotated to the credentials of the backup. this is so that operations using mysqlsh or mysqlci work